File: 24.swap-nodes-in-pairs/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
	ListNode* swapPairs(ListNode* head) {
		if (!head || !head->next)
		{
			return head;
		}
		ListNode *p = head->next;
		head->next = head->next->next;
		p->next = head;
		head = p;
		p = p->next;
		ListNode *pp = p->next;
		while (pp)
		{
			ListNode *q = pp->next;
			if (!q)
				break;
			pp->next = q->next;
			q->next = pp;
			p->next = q;
			p = pp;
			pp = p->next;
		}
		return head;
	}
};

File: 24.swap-nodes-in-pairs/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        p = ListNode(0)
        p.next = head
        head = p
        while p.next is not None and p.next.next is not None:
            tmp = p.next
            p.next = tmp.next
            tmp.next = p.next.next
            p.next.next = tmp
            p = p.next.next
        return head.next
